packages/rpc/json2/server.go

```diff
@@ -0,0 +1,259 @@
+
+// newCodecRequest returns a new CodecRequest.
+func newCodecRequest(r *http.Request, encoder rpc.Encoder, errorMapper func(error) error) rpc.CodecRequest {
+	// Decode the request body and check if RPC method is valid.
+	req := new(serverRequest)
+	err := json.NewDecoder(r.Body).Decode(req)
+
+	if err != nil {
+		err = &json2.Error{
+			Code:    json2.E_PARSE,
+			Message: err.Error(),
+			Data:    req,
+		}
+	} else if req.Version != version {
+		err = &json2.Error{
+			Code:    json2.E_INVALID_REQ,
+			Message: "jsonrpc must be " + version,
+			Data:    req,
+		}
+	}
+
+	_ = r.Body.Close()
+	return &CodecRequest{request: req, err: err, encoder: encoder, errorMapper: errorMapper}
+}
+
+// CodecRequest decodes and encodes a single request.
+type CodecRequest struct {
+	request     *serverRequest
+	err         error
+	encoder     rpc.Encoder
+	errorMapper func(error) error
+}
+
+// Method returns the RPC method for the current request.
+//
+// The method uses a dotted notation as in "Service.Method".
+func (c *CodecRequest) Method() (string, error) {
+	if c.err == nil {
+		return c.request.Method, nil
+	}
+	return "", c.err
+}
+
+// ReadRequest fills the request object for the RPC method.
+//
+// ReadRequest parses request parameters in two supported forms in
+// accordance with http://www.jsonrpc.org/specification#parameter_structures
+//
+// by-position: params MUST be an Array, containing the
+// values in the Server expected order.
+//
+// by-name: params MUST be an Object, with member names
+// that match the Server expected parameter names. The
+// absence of expected names MAY result in an error being
+// generated. The names MUST match exactly, including
+// case, to the method's expected parameters.
+func (c *CodecRequest) ReadRequest(args interface{}) error {
+	if c.err == nil && c.request.Params != nil {
+		// Note: if c.request.Params is nil it's not an error, it's an optional member.
+		// JSON params structured object. Unmarshal to the args object.
+		if err := json.Unmarshal(*c.request.Params, args); err != nil {
+			// Clearly JSON params is not a structured object,
+			// fallback and attempt an unmarshal with JSON params as
+			// array value and RPC params is struct. Unmarshal into
+			// array containing the request struct.
+			var params []interface{}
+			argsVal := reflect.Indirect(reflect.ValueOf(args))
+			// Parses unstructured JSON to Struct.
+			// Example:
+			//	struct Notification {
+			//		Message  string
+			//		Priority uint8
+			//		Critical bool
+			//	}
+			//	Unstructured JSON: ["Hello world", 10, false]
+			//	After parsing to Notification Struct: main.Notification{Message:"Hello world", Priority:0xa, Critical:false}
+			if argsVal.Kind() == reflect.Struct && argsVal.NumField() > 0 {
+				for i := 0; i < argsVal.NumField(); i++ {
+					params = append(params, argsVal.Field(i).Addr().Interface())
+				}
+			} else {
+				params = []interface{}{args}
+			}
+
```

Comment:
This is great, nice to parse these into structs here so the rpc methods can deal with structs instead of generic interfaces.  I like it!